### PR TITLE
feat: (Platform) Select component creation

### LINKED ISSUE #2521: platform: Select component creation
#### Is this a bug, enhancement, or feature request?

feature
#### Briefly describe your proposal.

Create select component in platform by wrapping up core.

#### Is there anything else we should know?

Re-tests this with FormGroup -> FormField setup.

### PR DESCRIPTION
**Pull request by sKudum**
#### Please provide a link to the associated issue.
closes #2521

#### Please provide a brief summary of this pull request.

1. Select component creation in platform which supports datasource and  platform form.
2. Few accessibility changes addressed in **core** select as suggested by QA team for the purpose of testing.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [n/a ] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Comment by sKudum** [comment]
netify url is not updated with latest changes, giving try without WIP in title

File: libs/platform/src/lib/components/form/select/select/select.component.ts [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+@Component({
+    selector: 'fdp-select',
+    templateUrl: './select.component.html',
+    styleUrls: ['./select.component.scss'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    encapsulation: ViewEncapsulation.None,
+    providers: [{ provide: FormFieldControl, useExisting: SelectComponent, multi: true }]
+})
+export class SelectComponent extends BaseSelect implements OnInit, AfterViewInit, AfterViewChecked {
```
Comment (fkolar) [inline libs/platform/src/lib/components/form/select/select/select.component.ts:39]: I am not sure if select shoudl have DataSource. DataSource are for components that requires to handle larger datasets. and for Select we had List binding

File: apps/docs/src/app/platform/platform-documentation.routes.ts [hunk 2]
```diff
@@ -94,8 +94,8 @@ export const ROUTES: Routes = [
             {
                 path: 'select',
                 loadChildren: () =>
```
Comment (fkolar) [inline apps/docs/src/app/platform/platform-documentation.routes.ts:96]: Please add example also for noSelectionValue.

File: libs/platform/src/lib/components/form/select/commons/base-select.ts [hunk 3]
```diff
@@ -0,0 +1,759 @@
...
+    @Input()
+    maxHeight = '250px';
+
+    /**
+     * Todo: Name of the entity for which DataProvider will be loaded. You can either pass list of
+     * items or use this entityClass and internally we should be able to do lookup to some registry
+     * and retrieve the best matching DataProvider that is set on application level
+     */
+    @Input()
+    entityClass: string;
```
Comment (fkolar) [inline libs/platform/src/lib/components/form/select/commons/base-select.ts:83]: EntityClass is only valid for datasources . We dont have usecase for DS yet for this component

File: libs/platform/src/lib/components/form/select/commons/base-select.ts [hunk 4]
```diff
@@ -0,0 +1,759 @@
...
+
+import { isFunction, isJsObject, isString } from '../../../../utils/lang';
+import { CollectionBaseInput } from '../../collection-base.input';
+import { SelectComponent } from '../select/select.component';
+import { SelectConfig, MatchingStrategy } from '../select.config';
+import { ContentDensity, FormFieldControl } from '../../form-control';
+import { FormField } from '../../form-field';
+import { TextAlignment } from '../../combobox';
+
+export type FdpSelectDataSource<T> = SelectDataSource<T> | Observable<T[]> | T[];
```
Comment (fkolar) [inline libs/platform/src/lib/components/form/select/commons/base-select.ts:62]: Please see my commend for datasource

File: libs/platform/src/lib/components/form/select/commons/base-select.ts [hunk 5]
```diff
@@ -0,0 +1,759 @@
...
+
+export class FdpSelectionChangeEvent {
+    constructor(
+        public source: SelectComponent,
+        public payload: any // Contains selected item
+    ) {}
+}
+
+@Directive()
+export abstract class BaseSelect extends CollectionBaseInput implements AfterViewInit, OnDestroy {
```
Comment (fkolar) [inline libs/platform/src/lib/components/form/select/commons/base-select.ts:72]: CollectionBaseInput should have a list binding

File: libs/platform/src/lib/components/form/select/select/select.component.ts [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+    @Input()
+    typeaheadDebounceInterval = 250;
+
+    /** Binds to control aria-labelledBy attribute */
+    @Input()
+    ariaLabelledBy: string = null;
+
+    /** Sets control aria-label attribute value */
+    @Input()
+    ariaLabel: string = null;
```
Comment (fkolar) [inline libs/platform/src/lib/components/form/select/select/select.component.ts:125]: Not sure why you have several properties defined both in this class and parent class

**Review by fkolar — CHANGES_REQUESTED** [review]

**Comment by sKudum** [comment]
Hi @Frank,

1).DataSource is removed used list binding of collectionBaseInput.
2).Removed unused code.
3).Added more examples including for no selection.
4).Add test cases testing more bindings with form-controls.


Few Points:
1).As per fiori spec select should support 2 columns, but no grouping(belongs to combobox) with flexbile column ratio- updated that feature as well.

2). Platform select is wrapper around core select and core select looks for fd-option component  as child element.by this not able to achieve template driven approach in platform. 

3).Select should support text clipping after 2/3 rows. did not find any disruption of it in  git book neither in fundamental-styles.- Are we planing to have to in feature once styles are ready?.

File: libs/platform/src/lib/components/form/select/commons/base-select.ts [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+    @Input()
+    set contentDensity(contentDensity: ContentDensity) {
+        this._contentDensity = contentDensity;
+        this.isCompact = contentDensity === 'compact';
+    }
+
+    /** Data for suggestion list */
+    @Input()
+    get list(): any {
+        return this._listData;
```
Comment (fkolar) [inline libs/platform/src/lib/components/form/select/commons/base-select.ts:211]: Have you checked the parent class for existing list Input?

**Review by fkolar — CHANGES_REQUESTED** [review]
The rest looks good to me , we just need to reuse what is in parent classes

**Review by KevinOkamoto — CHANGES_REQUESTED** [review]

**Review by fkolar — APPROVED** [review]
Looks good

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html [hunk 8]
```diff
@@ -0,0 +1,139 @@
+<fd-docs-section-title [id]="'mode'" [componentName]="'select'">
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:1]: unnecessary binding

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+     ratio for each columns by default ratio is 3:2 which is equal to 60%:40%.
+</description>
+<component-example>
+    <fdp-select-columns-example></fdp-select-columns-example>
+</component-example>
+<code-example [exampleFiles]="selectColumns"></code-example>
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'semantic'" [componentName]="'select'">
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:28]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    Use semantic states to visually highlight current <code>state</code> of the Select.
+</description>
+<component-example>
+    <fdp-select-semantic-state-example></fdp-select-semantic-state-example>
+</component-example>
+<code-example [exampleFiles]="selectSemantic"></code-example>
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'custom-select-template'" [componentName]="'select'">
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:41]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+</description>
+<component-example>
+    <fdp-select-custom-trigger></fdp-select-custom-trigger>
+</component-example>
+<code-example [exampleFiles]="customSelectTemplate"></code-example>
+
+<separator></separator>
+
+
+<fd-docs-section-title [id]="'select-mobile'" [componentName]="'select'">
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:62]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    Check <a [routerLink]="['/core/global-config']" fragment="mobile-mode-config">Global Configuration page</a> for more information.
+</description>
+<component-example>
+    <fdp-select-mobile-example></fdp-select-mobile-example>
+</component-example>
+<code-example [exampleFiles]="selectMobile"></code-example>
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'select-max-height'" [componentName]="'select'">
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:85]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+    window height.
+</description>
+<component-example>
+    <fdp-select-max-height-example></fdp-select-max-height-example>
+</component-example>
+<code-example [exampleFiles]="selectMaxHeight"></code-example>
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'select-none'" [componentName]="'select'">
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:101]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+    user to move to no value selected state.
+</description>
+<component-example>
+    <fdp-select-none-example></fdp-select-none-example>
+</component-example>
+<code-example [exampleFiles]="selectNone"></code-example>
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'select-nowrap'" [componentName]="'select'">
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:115]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    They get truncated with ellipsis.
+</description>
+<component-example>
+    <fdp-select-nowrap-example></fdp-select-nowrap-example>
+</component-example>
+<code-example [exampleFiles]="selectNoWrap"></code-example>
+
+<separator></separator>
+
+<fd-docs-section-title [id]="'reactive-form'" [componentName]="'select'">
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-docs.component.html:130]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-height/platform-select-max-height-example.component.html [hunk 16]
```diff
@@ -0,0 +1,8 @@
+<fdp-select 
+placeholder="Select an option"
+(selectionChange)="onSelect($event)"
+[maxHeight]="'250px'"
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-height/platform-select-max-height-example.component.html:4]: unnecessary binding

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-mode-example/platform-select-mode-example.component.ts [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        new Fruit('C', 'Plums', 10),
+        new Fruit('D', 'pineapple', 11),
+        new Fruit('E', 'watermelon', 10)
+    ];
+    option = this.userList.map<OptionItem>((item) => {
+        return {
+            label: item.name + item.id,
+            value: item,
+            triggerValue: '( ' + item.id + ' )',
+            disabled: item.id === 'B' ? true : false,
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-mode-example/platform-select-mode-example.component.ts:35]:
```suggestion
            triggerValue: `(${item.id})`,
            disabled: item.id === 'B',
```

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-none/platform-select-none-example.component.html [hunk 18]
```diff
@@ -0,0 +1,8 @@
+<fdp-select 
+placeholder="Select an option"
+width="7rem"
+(selectionChange)="onSelect($event)"
+[noValueLabel]="'None'"
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-none/platform-select-none-example.component.html:5]: unnecessary binding

File: libs/core/src/lib/select/option/option.component.ts [hunk 19]
```diff
@@ -24,6 +24,7 @@ import {
 
 import { KeyUtil } from '../../utils/functions/key-util';
 import { OptionsInterface } from '../options.interface';
+import { By } from '@angular/platform-browser';
```
Comment (Betrozov) [inline libs/core/src/lib/select/option/option.component.ts:27]: do you need it?

File: libs/platform/src/lib/components/form/select/select/select.component.html [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+[closeOnOutsideClick]="closeOnOutsideClick"
+[controlTemplate]="controlTemplate"
+[appendTo]="appendTo"
+[fillControlMode]="fillControlMode"
+[maxHeight]="maxHeight"
+[attr.aria-disabled]="disabled || readonly"
+[attr.aria-labelledby]="value"
+[attr.tooltip]="value"
+[attr.aria-readonly]="readonly"
+[attr.role]="'listbox'"
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/select/select/select.component.html:15]: unnecessary binding

File: libs/platform/src/lib/components/form/select/select/select.component.ts [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+        readonly _selectConfig: SelectConfig,
+        @Optional() @SkipSelf() @Host() formField: FormField,
+        @Optional() @SkipSelf() @Host() formControl: FormFieldControl<any>
+    ) {
+        super(cd, elementRef, ngControl, ngForm, _selectConfig, formField, formControl);
+    }
+
+    /** @hidden
+     */
+    ngOnInit(): void {}
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/select/select/select.component.ts:90]: please remove empty state

File: libs/platform/src/lib/components/form/select/select/select.component.ts [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+            if (this.maxWidth) {
+                optionItem.setAttribute('style', 'max-width: ' + this.maxWidth + 'rem');
+            }
+            // by default text will be overlapped, below it will help to truncate
+            if (this.noWrapText) {
+                optionItem.querySelector('.fd-list__title').setAttribute(
+                    'title', optionItem.querySelector('.fd-list__title').innerHTML);
+                optionItem.querySelector('.fd-list__title').setAttribute(
+                    'aria-label', optionItem.querySelector('.fd-list__title').innerHTML);
+                optionItem.querySelector('.fd-list__title').classList.add('fd-list__title--no-wrap');
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/select/select/select.component.ts:120]:
```suggestion
                const listTitle = optionItem.querySelector('.fd-list__title');
                const listTitleInnerHTML = optionItem.querySelector('.fd-list__title').innerHTML;
                listTitle.setAttribute('title', listTitleInnerHTML);
                listTitle.setAttribute('aria-label', listTitleInnerHTML);
                listTitle.classList.add('fd-list__title--no-wrap');
```

File: libs/platform/src/lib/components/form/select/select/select.component.ts [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+        return this.lookupKey && this.lookupValue(this.selected) === this.lookupValue(selectedItem) ||
+            this.displayValue(this.selected) === this.displayValue(selectedItem);
+    }
+
+    /**
+     * Define is selected item selected by display value
+     * @hidden
+     */
+    _isSelectedOptionItemByDisplayValue(selectedItem: any): boolean {
+        return this.selected && this.selected.label === selectedItem;
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/select/select/select.component.ts:213]:
```suggestion
        return this.selected?.label === selectedItem;
```

File: libs/platform/src/lib/components/form/select/commons/base-select.ts [hunk 24]
```diff
@@ -0,0 +1,551 @@
...
+    searchInputElement: ElementRef;
+
+    /** @hidden */
+    _contentDensity: ContentDensity = this.selectConfig.contentDensity;
+
+    /**
+      * @hidden
+      * Used to define if contentDensity value is 'compact' or not.
+    */
+   _isCompact = this._contentDensity === 'compact';
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/select/commons/base-select.ts:259]: seems it not used

**Review by Betrozov — CHANGES_REQUESTED** [review]

File: libs/platform/src/lib/components/form/select/select/select.component.ts (hunk 23 above)
Comment (sKudum) [reply libs/platform/src/lib/components/form/select/select/select.component.ts:213]: > 
Can't do this change as "Expression expected error "was getting displayed on compilation

File: libs/platform/src/lib/components/form/select/select/select.component.ts (hunk 21 above)
Comment (sKudum) [reply libs/platform/src/lib/components/form/select/select/select.component.ts:90]: removal of it leads to compile time error as =>"at SelectComponent.push.FWTC.BaseInput.ngOnInit"

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-semantic-state-example/platform-select-semantic-state-example.component.ts [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+        new Fruit('C', 'Plums', 10),
+        new Fruit('D', 'pineapple', 11),
+        new Fruit('E', 'watermelon', 10)
+    ];
+    option = this.userList.map<OptionItem>((item) => {
+        return {
+            label: item.name + item.id,
+            value: item,
+            triggerValue: `(${item.id})`,
+            disabled: item.id === 'B' ? true : false,
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-select/platform-select-examples/platform-select-semantic-state-example/platform-select-semantic-state-example.component.ts:35]:
```suggestion
            disabled: item.id === 'B',
```

File: libs/platform/src/lib/components/form/select/select/select.component.ts [hunk 26]
```diff
@@ -0,0 +1,255 @@
...
+        this.select._options.forEach((option) => {
+            const optionItem = option._getHtmlElement();
+            // set maxWidth default is 40rem
+            if (this.maxWidth) {
+                optionItem.setAttribute('style', 'max-width: ' + this.maxWidth + 'rem');
+            }
+            // by default text will be overlapped, below it will help to truncate
+            if (this.noWrapText) {
+                const listTitle = optionItem.querySelector('.fd-list__title');
+                const listTitleInnerHTML = optionItem.querySelector('.fd-list__title').innerHTML;
```
Comment (Betrozov) [inline libs/platform/src/lib/components/form/select/select/select.component.ts:118]:
```suggestion
                const listTitleInnerHTML = listTitle.innerHTML;
```

**Review by Betrozov — CHANGES_REQUESTED** [review]

**Review by Betrozov — APPROVED** [review]

**Review by KevinOkamoto — APPROVED** [review]

**Comment by puru-hk** [comment]
In RTL mode UI is not proper attache screenshot

**Comment by puru-hk** [comment]
In RTL Mobile display is not proper:

image

**Comment by sKudum** [comment]
> In RTL Mobile display is not proper:
> 
> image
> 
> image

Hi @puru-hk ,

Issues are fixed , please have a look.

After fix:

**Comment by puru-hk** [comment]
Looks good

**Review by puru-hk — APPROVED** [review]

**Review by JKMarkowski — APPROVED** [review]
LGTM - Great job with fixing tab out issue

**Pull request merged**